apiserver/facades/client/modelgeneration/modelgeneration.go

```diff
@@ -127,23 +145,29 @@ func (m *ModelGenerationAPI) AdvanceGeneration(arg params.AdvanceGenerationArg)
 		case names.UnitTagKind:
 			results.Results[i].Error = common.ServerError(generation.AssignUnit(tag.Id()))
 		default:
-			results.Results[i].Error = common.ServerError(errors.Errorf("expected names.UnitTag or names.ApplicationTag, got %T", tag))
+			results.Results[i].Error = common.ServerError(
+				errors.Errorf("expected names.UnitTag or names.ApplicationTag, got %T", tag))
 		}
-		err = generation.Refresh()
-		if err != nil {
-			return results, errors.Trace(err)
+		if err := generation.Refresh(); err != nil {
+			return params.AdvanceGenerationResult{AdvanceResults: results}, errors.Trace(err)
 		}
 	}
 
+	result := params.AdvanceGenerationResult{AdvanceResults: results}
+
 	ok, err := generation.CanAutoComplete()
 	if err != nil {
-		return results, errors.Trace(err)
+		return result, errors.Trace(err)
```

Comment:
Maybe annotate and set in `CompleteResult.Error`?

Reply:
What kind of errors can happen here? It feels like this is truly exceptional (db crashed, etc), so having it be an error-to-the-request is ok. Certainly it doesn't have to do with the individual request entities.

That said, it does feel a bit odd that AutoComplete sets CompleteResult, but CanAutoComplete sets the global, but I think its ok given the types of errors. (Presumably AutoComplete can fail because things aren't in consistent state?)

Is there a race here, where you check CanAutoComplete, and in the meantime someone adds another application to the generation, and then AutoComplete would accidentally complete that application? Or would you just "unable to autocomplete" at that point?
Would it be better to avoid a "Can" and just always call "AutoCompleteIfPossible" to avoid the race?